### PR TITLE
Fixes #32345 - Provide non-DB model methods for FactParser

### PR DESCRIPTION
**Pull request by ekohl**
As a preparation for removel of the DB model methods, there should be an alternative for DB model methods (Architecture, Domain, Environment & Model) while also deprecating the DB model methods.

Currently it doesn't have the deprecations and no code has been modified to use the new methods, but it's to split off some part of https://github.com/theforeman/foreman/pull/8449 to make that PR smaller.

### CODE REVIEW

File: app/services/fact_parser.rb [hunk 1]
```diff
@@ -61,10 +81,14 @@ def has_comment?
   end
 
   def hostgroup
-    hostgroup_title = facts[:foreman_hostgroup]
+    hostgroup_title = hostgroup_name
     Hostgroup.unscoped.where(:title => hostgroup_title).first_or_create if hostgroup_title.present?
   end
 
+  def hostgroup_name
```
Comment (ekohl) [inline app/services/fact_parser.rb:88]: @ofedoren suggested this in the related PR:
```suggestion
  def hostgroup_title
```

I'm not sure which is best, but it shouldn't be lost.

File: app/services/fact_parser.rb [hunk 2]
```diff
@@ -33,18 +33,38 @@ def operatingsystem
...
+    Model.where(:name => name.strip).first_or_create if name.present?
+  end
+
+  def model_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def domain
+    name = domain_name
+    Domain.unscoped.where(:name => name).first_or_create if name.present?
```
Comment (ekohl) [inline app/services/fact_parser.rb:64]: Prior discussion: https://github.com/theforeman/foreman/pull/8449#discussion_r614895291

tl;dr: unscoped is a risk, but it was in the Puppet fact parser where I copied it from. [foreman_ansible](https://github.com/theforeman/foreman_ansible/blob/5eec05f645eae1b1cde9de2179f4cf5208873ae3/app/services/foreman_ansible/fact_parser.rb#L28-L32) doesn't use unscoped. There may be more out there.

Comment (lzap) [reply app/services/fact_parser.rb:64]: Yes this is a long-standing issue (also for discovery). New hosts are even assigned organization and location via foreman_organization administer setting. I think the correct solution would be a boolean flag for each resource (domain, subnet) named "allow discovery of hosts via facts". Users would need to explicitly enable or disable those which are whitelisted.

File: app/services/fact_parser.rb [hunk 3]
```diff
@@ -33,18 +33,38 @@ def operatingsystem
   end
 
   def environment
+    name = environment_name
+    Environment.unscoped.where(:name => name).first_or_create
```
Comment (lzap) [inline app/services/fact_parser.rb:37]: Nitpick: You can use new hash syntax.

File: app/services/fact_parser.rb (hunk 1 above)
Comment (lzap) [reply app/services/fact_parser.rb:88]: Well title seems better fit for me.

**Review by lzap — CHANGES_REQUESTED** [review]
It should be really "title" because "name" is confusing.

Discovery does work with this change, no need to patch it. Other than that I am good.

**Comment by lzap** [comment]
The Ansible patch is RTM, let me know once you want me to re-review this.
